cdc/kv/client_test.go

```diff
@@ -61,6 +70,128 @@ func (s *clientSuite) TestUpdateCheckpointTS(c *check.C) {
 	c.Assert(checkpointTS, check.Equals, maxValue)
 }
 
+type mockChangeDataService struct {
+	c  *check.C
+	ch chan *cdcpb.ChangeDataEvent
+}
+
+func (s *mockChangeDataService) EventFeed(server cdcpb.ChangeData_EventFeedServer) error {
+	req, err := server.Recv()
+	s.c.Assert(err, check.IsNil)
+	initialized := &cdcpb.ChangeDataEvent{
+		Events: []*cdcpb.Event{
+			{
+				RegionId:  req.RegionId,
+				RequestId: req.RequestId,
+				Event: &cdcpb.Event_Entries_{
+					Entries: &cdcpb.Event_Entries{
+						Entries: []*cdcpb.Event_Row{
+							{
+								Type: cdcpb.Event_INITIALIZED,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	err = server.Send(initialized)
+	s.c.Assert(err, check.IsNil)
+	for e := range s.ch {
+		for _, event := range e.Events {
+			event.RequestId = req.RequestId
+		}
+		err := server.Send(e)
+		s.c.Assert(err, check.IsNil)
+	}
+	return nil
+}
+
+func newMockService(c *check.C, port int, ch chan *cdcpb.ChangeDataEvent) *grpc.Server {
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	c.Assert(err, check.IsNil)
+	grpcServer := grpc.NewServer()
+	mockService := &mockChangeDataService{c: c, ch: ch}
+	cdcpb.RegisterChangeDataServer(grpcServer, mockService)
+	go func() {
+		err := grpcServer.Serve(lis)
+		c.Assert(err, check.IsNil)
+	}()
+	return grpcServer
+}
+
+// Use etcdSuite to workaround the race. See comments of `TestConnArray`.
+func (s *etcdSuite) TestConnectOfflineTiKV(c *check.C) {
+	ch2 := make(chan *cdcpb.ChangeDataEvent, 10)
+	defer func() {
+		if ch2 != nil {
+			close(ch2)
+		}
+	}()
+	server2 := newMockService(c, 23376, ch2)
+	defer func() {
+		// if server1 != nil {
+		// 	server1.Stop()
+		// }
+		if server2 != nil {
+			server2.Stop()
+		}
+	}()
+
+	cluster := mocktikv.NewCluster()
+	rpcClient, pdClient, err := mocktikv.NewTiKVAndPDClient(cluster, nil, "")
+	c.Assert(err, check.IsNil)
+	kvStorage, err := tikv.NewTestTiKVStore(rpcClient, pdClient, nil, nil, 0)
+	c.Assert(err, check.IsNil)
+
+	cluster.AddStore(1, "localhost:23375")
+	cluster.AddStore(2, "localhost:23376")
+	cluster.Bootstrap(3, []uint64{1, 2}, []uint64{4, 5}, 4)
+
+	cdcClient, err := NewCDCClient(pdClient, kvStorage.(tikv.Storage))
+	c.Assert(err, check.IsNil)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	eventCh := make(chan *model.RegionFeedEvent, 10)
+	go func() {
+		err := cdcClient.EventFeed(ctx, util.Span{Start: []byte("a"), End: []byte("b")}, 1, eventCh)
```

Comment:
`cdcClient.EventFeed` may not return before the test case tear down, what about adding some wait mechanism